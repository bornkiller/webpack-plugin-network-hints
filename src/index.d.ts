export interface NetworkHintsOptions {
  prefetch?: string[];
  preload?: string[];
}

export class NetworkHintsPlugin {
  constructor(options: NetworkHintsOptions) {}
}
