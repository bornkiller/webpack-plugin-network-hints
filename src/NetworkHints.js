/**
 * @description - webpack plugin for inject network hints
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// packages
import { resolve } from 'url';
import { match } from 'minimatch';
import HtmlWebpackPlugin from 'html-webpack-plugin';

// internal
import RequireHtmlWebpackPluginError from './exceptions/RequireHtmlWebpackPluginError';
import UnsupportHtmlWebpackPluginVersionError from './exceptions/UnsupportHtmlWebpackPluginVersionError';
import preloadAttrAlgorithm from './preload';

// scope
const defaultOptions = {
  prefetch: [],
  preload: [],
};

class NetworkHints {
  constructor(options) {
    this.options = { ...defaultOptions, ...options };
  }

  apply(compiler) {
    compiler.hooks.compilation.tap(
      'WebpackPluginNetworkHints',
      (compilation) => {
        const [HtmlWebpackPluginInstance] = compiler.options.plugins.filter(
          (plugin) => plugin.constructor.name === 'HtmlWebpackPlugin'
        );

        if (!HtmlWebpackPluginInstance) {
          compilation.errors.push(new RequireHtmlWebpackPluginError());
        }
        // pre-v4 HtmlWebpackPlugin
        else if (
          Reflect.has(compilation.hooks, 'htmlWebpackPluginAlterAssetTags')
        ) {
          compilation.errors.push(new UnsupportHtmlWebpackPluginVersionError());
        } else {
          // get available hooks directly
          const hooks = HtmlWebpackPlugin.getHooks(compilation);

          hooks.alterAssetTagGroups.tapAsync(
            'InjectExternalPlugin',
            (assetTagGroups, callback) => {
              const { publicPath } = compiler.options.output;
              const assets = Reflect.ownKeys(compilation.assets);
              const prefetch = this.options.prefetch
                .reduce(
                  (acc, pattern) => [
                    ...acc,
                    ...match(assets, pattern, { matchBase: true }),
                  ],
                  []
                )
                .sort()
                .map((asset) => resolve(publicPath, asset))
                .map((asset) =>
                  HtmlWebpackPlugin.createHtmlTagObject('link', {
                    rel: 'prefetch',
                    href: asset,
                  })
                );
              const preload = this.options.preload
                .reduce(
                  (acc, pattern) => [
                    ...acc,
                    ...match(assets, pattern, { matchBase: true }),
                  ],
                  []
                )
                .sort() // consistant assets order
                .map((asset) => resolve(publicPath, asset))
                .map((asset) =>
                  HtmlWebpackPlugin.createHtmlTagObject(
                    'link',
                    preloadAttrAlgorithm(asset)
                  )
                );

              const payload = {
                ...assetTagGroups,
                headTags: [...assetTagGroups.headTags, ...prefetch, ...preload],
              };

              callback(null, payload);
            }
          );
        }
      }
    );
  }
}

export default NetworkHints;
