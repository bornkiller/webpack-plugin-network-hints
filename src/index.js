/**
 * @description - nothing but re-export
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// eslint-disable-next-line import/prefer-default-export
export { default as NetworkHintsPlugin } from './NetworkHints';
