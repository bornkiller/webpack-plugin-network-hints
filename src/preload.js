/**
 * @description - calculate preload link attributes
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// packages
import { extname } from 'path';
import { lookup } from 'mime-types';

// scope
const scripts = ['.js', '.mjs'];
const stylesheets = ['.css'];
const fonts = ['.woff', '.woff2', '.ttf', '.eot'];
const images = ['.jpg', '.jpeg', '.png', '.gif', '.webp'];
const documents = ['.html'];

/**
 * @description - compose specific attrs for specific files
 *
 * @param {string} asset
 */
function preloadSpecificAttrs(asset) {
  const extension = extname(asset);

  switch (true) {
    case scripts.includes(extension):
      return { as: 'script' };
    case stylesheets.includes(extension):
      return { as: 'stylesheet', onload: `this.rel='stylesheet'` };
    case fonts.includes(extension):
      return { as: 'font', crossorigin: 'anonymous' };
    case images.includes(extension):
      return { as: 'image' };
    case documents.includes(extension):
      return { as: 'document' };
    default:
      return {};
  }
}

/**
 * @param {string} asset
 *
 * @return {object}
 */
export default function preload(asset) {
  const attrs = {
    rel: 'preload',
    href: asset,
    type: lookup(asset),
  };
  const specifics = preloadSpecificAttrs(asset);

  return { ...attrs, ...specifics };
}
