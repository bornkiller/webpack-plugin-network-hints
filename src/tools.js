/**
 * @description - tool method
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

/* eslint-disable import/prefer-default-export */

/**
 * @description - create prefetch tags
 *
 * @param {string} asset
 */
export function createPrefetchTags(asset) {
  return {
    tagName: 'link',
    selfClosingTag: false,
    voidTag: true,
    attributes: {
      rel: 'prefetch',
      href: asset,
    },
  };
}

/**
 * @description - create preload tags
 *
 * @param {string} asset
 * @param {function} algorithm - calculate preload additional attributes
 */
export function createPreloadTags(asset, algorithm) {
  return {
    tagName: 'link',
    selfClosingTag: false,
    voidTag: true,
    attributes: algorithm(asset),
  };
}
