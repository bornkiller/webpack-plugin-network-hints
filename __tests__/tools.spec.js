/**
 * @description - tools unit tests
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// packages
import { createPrefetchTags, createPreloadTags } from '../src/tools';

// scope
const algorithm = jest
  .fn()
  .mockReturnValue({ as: 'font', crossorigin: 'anonymous' });

describe('tools suits', () => {
  it('should create prefetch tag', () => {
    expect(createPrefetchTags('/hello.jpg')).toMatchSnapshot();
  });

  it('should create preload tag', () => {
    expect(createPreloadTags('/iconfont.woff2', algorithm)).toMatchSnapshot();
  });
});
