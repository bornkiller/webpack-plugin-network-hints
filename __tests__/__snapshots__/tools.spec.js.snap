// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tools suits should create prefetch tag 1`] = `
Object {
  "attributes": Object {
    "href": "/hello.jpg",
    "rel": "prefetch",
  },
  "selfClosingTag": false,
  "tagName": "link",
  "voidTag": true,
}
`;

exports[`tools suits should create preload tag 1`] = `
Object {
  "attributes": Object {
    "as": "font",
    "crossorigin": "anonymous",
  },
  "selfClosingTag": false,
  "tagName": "link",
  "voidTag": true,
}
`;
